core/near-network/src/lib.rs

```diff
@@ -0,0 +1,439 @@
+    listener: Option<TcpListener>,
+    /// Local info about accounts
+    account_to_peer: Arc<RwLock<HashMap<AccountId, PeerId>>>,
+    /// connected peers and the sender channel to the peer
+    connected_peers: Arc<RwLock<HashMap<PeerId, Sender<ServiceEvent>>>>,
+    /// state of the connection with peer
+    peer_state: Arc<RwLock<HashMap<SocketAddr, ConnectionState>>>,
+    /// PeerId to peer info
+    peer_info: Arc<RwLock<HashMap<PeerId, Peer>>>,
+    /// number of peers to gossip
+    gossip_num: usize,
+    /// channel that sends custom message for further processing
+    message_tx: Sender<NetworkMessage>,
+}
+
+impl Service {
+    pub fn new(addr: &str, peer_id: PeerId, message_tx: Sender<NetworkMessage>) -> Self {
+        let addr = addr.parse::<SocketAddr>().expect("Incorrect address");
+        let listener = TcpListener::bind(&addr).expect("Cannot bind to address");
+        Service {
+            peer_id,
+            account_id: None,
+            listener: Some(listener),
+            account_to_peer: Arc::new(RwLock::new(HashMap::new())),
+            connected_peers: Arc::new(RwLock::new(HashMap::new())),
+            peer_state: Arc::new(RwLock::new(HashMap::new())),
+            peer_info: Arc::new(RwLock::new(HashMap::new())),
+            gossip_num: 3,
+            message_tx
+        }
+    }
+
+    pub fn new_with_account_id(
+        addr: &str, 
+        peer_id: PeerId,
+        message_tx: Sender<NetworkMessage>,
+        account_id: AccountId,
+    ) -> Self {
+        let mut service = Self::new(addr, peer_id, message_tx);
+        service.account_id = Some(account_id);
+        service
+    }
+
+    fn get_connection_handler(&self) -> ConnectionHandler {
+        ConnectionHandler {
+            account_to_peer: self.account_to_peer.clone(),
+            connected_peers: self.connected_peers.clone(),
+            peer_state: self.peer_state.clone(),
+            peer_info: self.peer_info.clone(),
+            message_tx: self.message_tx.clone(),
+        }
+    }
+
+    /// try to dial peer, if we are already connected to the peer or are waiting to connect,
+    /// do nothing. Otherwise we create a future that represents the connection and put it
+    /// to the pending connection queue
+    pub fn dial(&self, addr: SocketAddr) {
+        if self.peer_state.read().contains_key(&addr) {
+            return;
+        }
+        self.peer_state.write().insert(addr, ConnectionState::Pending);
+        let connection_handler = self.get_connection_handler();
+        let peer_id = self.peer_id;
+        let account_id = self.account_id.clone();
+        let task = TcpStream::connect(&addr).map(move |stream| {
+            let sender = connection_handler.convert_stream(addr, stream);
+            Self::send_handshake_message(sender, peer_id, account_id);
+        }).map_err(|_| ());
+        tokio::spawn(task);
+    }
+
+    pub fn spawn_listening_task(&mut self) {
+        let peer_state = self.peer_state.clone();
```

Comment:
I was thinking of doing that and spawn all background tasks at once, but then call to `new` has to be inside a task, unless we spawn a thread from within `new`. Do we still want to access service after initialization?

Reply:
Ideally we don't want other parts of our infra to communicate with `Service` object directly. It should be done through the channels, for better compatibility with async code. Nothing outside network should be calling `dial` or `send_message` directly. Instead `dial` should be happening upon the spawning of the network task and other subtasks.

Reply:
Unit tests that are located in the same file have access to private methods and fields. Leave the current methods as they currently are but make them private, then continue using them in unit tests just like you are using it now. The code that is currently in the `new` method can be renamed to `init` and invoked from the `new` method before the tasks are spawned. Unit tests would then use private `init` method, while external code would use `new`.

The problem is that `dial` and `send_message` will be wrapped into tasks that expose channels anyway, because that's the only way to call them from code that uses tokio. This wrapping can be either done outside the network code or inside the network code. It makes sense to do it inside because this wrapping is network-specific.